libnavigation-core/src/main/java/com/mapbox/navigation/core/replay/history/ReplayRouteDriver.kt

```diff
@@ -0,0 +1,193 @@
+            )
+
+            val direction = TurfMeasurement.bearing(segmentStartPoint, segmentEndPoint)
+            for (i in 1 until segment.numElements - 1) {
+                val point = TurfMeasurement.destination(segmentStartPoint, segment.distances[i], direction, TurfConstants.UNIT_METERS)
+                val location = toBlankLocation(point)
+                location.time = segment.times[i].toLong()
+                location.speed = segment.speeds[i].toFloat()
+                gpsLocations.add(location)
+            }
+            timeMillis = segment.times.last().toLong()
+            segmentEnd.time = timeMillis
+            timeMillis += timeMillisIncrement
+            gpsLocations.add(segmentEnd)
+        }
+
+        createBearingProfile(smoothPoints, gpsLocations)
+
+        return gpsLocations
+    }
+
+    private fun createSpeedProfile(smoothPoints: List<Point>, smoothLocations: List<Location>) {
+        var bearing = TurfMeasurement.bearing(smoothPoints[0], smoothPoints[1])
+        smoothLocations.last().speed = 0.0f
+        smoothLocations.last().accuracy = 3.0f
+        smoothLocations.last().bearing = bearing.toFloat()
+        for (i in 1 until smoothLocations.size-1) {
+            val nextBearing = TurfMeasurement.bearing(smoothPoints[i], smoothPoints[i+1])
+            val deltaBearing = abs(nextBearing - bearing)
+            bearing = nextBearing
+
+            val velocityFraction = 1.0 - min(1.0, deltaBearing / 90.0)
+            val offsetToMaxVelocity = maxVelocity - minVelocityMps
+            smoothLocations[i].speed = (minVelocityMps + (velocityFraction * offsetToMaxVelocity)).toFloat()
+            smoothLocations[i].accuracy = 3.0f
+            smoothLocations[i].bearing = bearing.toFloat()
+            println("delta bearing $deltaBearing speed=${smoothLocations[i].speed}")
+            Log.i("RouteReplay", "RouteReplay deltaBearing=$deltaBearing speed=${smoothLocations[i].speed}")
+        }
+        // Stop at the end of the route
+        smoothLocations.last().speed = 0.0f
+        smoothLocations.last().accuracy = 3.0f
+        smoothLocations.last().bearing = bearing.toFloat()
+    }
+
+    private fun createBearingProfile(smoothPoints: List<Point>, gpsLocations: List<Location>) {
+        // TODO Create a bearing profile for the route
+        var bearing = TurfMeasurement.bearing(smoothPoints[0], smoothPoints[1])
+        gpsLocations.forEachIndexed { index, location ->
+            if (index + 3 < gpsLocations.lastIndex) {
+                val fromPoint = Point.fromLngLat(location.longitude, location.latitude)
+                val toPoint = Point.fromLngLat(gpsLocations[index + 3].longitude, gpsLocations[index + 3].latitude)
+                bearing = TurfMeasurement.bearing(fromPoint, toPoint)
+            }
+            location.bearing = bearing.toFloat()
+        }
+    }
+
+    private fun driveSegment(startTime: Int, startSpeed: Int, endSpeed: Int, distance: Double): DriveSegment {
```

Comment:
Here is a jupyter notebook used to figure this out
https://github.com/kmadsen/android/blob/master/positioning/interpolation/Intepolate-points.pdf